pkg/agent/proxy/proxier.go

```diff
@@ -218,59 +254,81 @@ func (p *proxier) installServices() {
 			addedLoadBalancerIPs = svcInfo.LoadBalancerIPStrings()
 		}
 		if len(deletedLoadBalancerIPs) > 0 || len(addedLoadBalancerIPs) > 0 {
-			needUpdate = true
+			needUpdateService = true
 		}
 
-		if !needUpdate {
+		// If neither the Service nor Endpoints of the Service need to be updated, we skip.
+		if !needUpdateService && !needUpdateEndpoints {
 			continue
 		}
 
-		if err := p.ofClient.InstallEndpointFlows(svcInfo.OFProtocol, endpointUpdateList, p.isIPv6); err != nil {
-			klog.Errorf("Error when installing Endpoints flows: %v", err)
-			continue
-		}
-		err := p.ofClient.InstallServiceGroup(groupID, svcInfo.StickyMaxAgeSeconds() != 0, endpointUpdateList)
-		if err != nil {
-			klog.Errorf("Error when installing Endpoints groups: %v", err)
-			p.endpointInstalledMap[svcPortName] = nil
-			continue
-		}
-		// Delete previous flow.
-		if needRemoval {
-			if err := p.ofClient.UninstallServiceFlows(pSvcInfo.ClusterIP(), uint16(pSvcInfo.Port()), pSvcInfo.OFProtocol); err != nil {
-				klog.Errorf("Failed to remove flows of Service %v: %v", svcPortName, err)
-			}
-		}
-		if err := p.ofClient.InstallServiceFlows(groupID, svcInfo.ClusterIP(), uint16(svcInfo.Port()), svcInfo.OFProtocol, uint16(svcInfo.StickyMaxAgeSeconds())); err != nil {
-			klog.Errorf("Error when installing Service flows: %v", err)
-			continue
-		}
-		// Install OpenFlow entries for the ingress IPs of LoadBalancer Service.
-		// The LoadBalancer Service should be accessible from Pod, Node and
-		// external host.
-		var toDelete, toAdd []string
-		if needRemoval {
-			toDelete = pSvcInfo.LoadBalancerIPStrings()
-			toAdd = svcInfo.LoadBalancerIPStrings()
+		if pSvcInfo != nil {
+			klog.V(2).Infof("Updating Service %s %s", svcPortName.Name, svcInfo.String())
 		} else {
-			toDelete = deletedLoadBalancerIPs
-			toAdd = addedLoadBalancerIPs
+			klog.V(2).Infof("Installing Service %s %s", svcPortName.Name, svcInfo.String())
 		}
-		for _, ingress := range toDelete {
-			if ingress != "" {
-				// It is safe to access pSvcInfo here. If this is a new Service,
-				// then toDelete will be an empty slice.
-				if err := p.uninstallLoadBalancerServiceFlows(net.ParseIP(ingress), uint16(pSvcInfo.Port()), pSvcInfo.OFProtocol); err != nil {
-					klog.Errorf("Error when removing LoadBalancer Service flows: %v", err)
-					continue
+
+		if needUpdateEndpoints {
+			err := p.ofClient.InstallEndpointFlows(svcInfo.OFProtocol, endpointUpdateList, p.isIPv6)
+			if err != nil {
```

Comment:
So, how we resync for these errors? Do we need to wait for the next Service or Endpoint update?

Reply:
The "sync" (`syncProxyRules` method) happens every 30s: https://github.com/vmware-tanzu/antrea/blob/ec1839765432cc97196431c676897e56afc74efd/pkg/agent/proxy/proxier.go#L448. We look for mismatches and handle them. By the way @weiqiangt should we adjust the constructor parameters to handle bursts better? I notice that upstream kube-proxy uses a min sync interval of 1s.

Reply:
Thanks for the explanation.Yes, 1s is more reasonable. Updated.

Reply:
So, kube-proxy also handles such errors by sync every 30s? And for a new change it will just trigger re-sync, after 1s?

The K8s controller pattern of exponential  backoff for the exact failed resources sounds a better strategy to me.

Reply:
@jianjuns,  I've upated the code to make the retry faster.
If there is a failure, we notify the runner to run as soon as possible, and the runner guarantees the sync won't be too frequently according to its minInterval.

Reply:
So, the resync will be just 1s, or there is some backoff? Do you know the kube-proxy behavior?

Reply:
No back-off for now since we rely on the runner to schedule the sync.
I think there are two ways to implement:
* We trigger the runner with backoff
We have a separate goroutine to catch any error and trigger the runner with backoff.
* We implement our own Runner based on the `BoundedFrequencyRunner` and bring the ability to retry with backoff to it.

AFAIK, I haven't found there is backoff strategy when the iptables-restore failed. But kube-proxy does have a fixed retry duration.

Reply:
To confirm again kube-proxy will retry after 1s? I am thinking if we are not sure, maybe let us use a conservative min interval for now, like 10s, as no time for scale tests.

Reply:
The default retry duration of kube-proxy is 1s which means the retry will occur with 1s. For the time 10s, I think it may be too long?

Reply:
This one second just for failure retry or also for other service/endpoint changes? If the latter, then let us go 1s.

Reply:
Got it. The default sync duration is also 1s, then lets keep current change.